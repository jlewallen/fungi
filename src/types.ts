export class StationNavigation {
    constructor(public readonly deviceId: string, public readonly name: string) {}

    public static fromRegistration(registration: Registration): StationNavigation {
        return new StationNavigation(registration.deviceId, registration.address);
    }
}

export class Registration {
    public readonly address: string;

    constructor(
        public readonly deviceId: string,
        public readonly addresses: string[],
        public readonly port: number,
        public zeroconf: Date | null = null,
        public udp: Date | null = null,
        public lost: Date | null = null,
        public queried: Date | null = null,
        public replied: Date | null = null,
        public refreshed: Date | null = null
    ) {
        this.address = this.addresses[0];
    }

    public clone(): Registration {
        return new Registration(
            this.deviceId,
            this.addresses,
            this.port,
            this.zeroconf,
            this.udp,
            this.lost,
            this.queried,
            this.replied,
            new Date()
        );
    }
}

export class PersistedStation {
    constructor(public readonly reply: AppProto.HttpReply, public readonly registration: Registration) {}
}
