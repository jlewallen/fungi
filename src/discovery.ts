import Zeroconf from "react-native-zeroconf";
import RNFetchBlob from "react-native-fetch-blob";

import dgram from "react-native-udp";
import { Buffer } from "buffer";
import { fk_app as AppProto } from "fk-app-protocol/fk-app";

export class Registration {
    constructor(
        public readonly name: string,
        public readonly addresses: string[],
        public readonly port: number,
        public zeroconf: Date | null = null,
        public udp: Date | null = null,
        public lost: Date | null = null,
        public queried: Date | null = null,
        public replied: Date | null = null
    ) {}
}

type CallbackFunc = (registrations: Registration[]) => Promise<void>;

export class Discovery {
    private readonly services: { [index: string]: Registration } = {};
    private callback: CallbackFunc = async () => {};
    private zeroconf: Zeroconf = null;

    async start(callback: CallbackFunc): Promise<void> {
        if (this.zeroconf) {
            return;
        }

        this.callback = callback;
        this.zeroconf = new Zeroconf();

        this.zeroconf.on("start", () => console.log("Zeroconf scan has started."));
        this.zeroconf.on("stop", () => console.log("Zeroconf scan has stopped."));
        this.zeroconf.on("found", (found: unknown) => console.log("Zeroconf found.", found));
        this.zeroconf.on("update", () => {
            // console.log('Zeroconf update.');
            this.refresh();
        });
        this.zeroconf.on("error", (error: string) => console.log("Zeroconf error.", error));

        this.zeroconf.on("resolved", (resolved: { name: string; addresses: string[]; port: number }) => {
            console.log("Zeroconf found.", resolved);
            this.onZeroConfFound(resolved.name, resolved.addresses, resolved.port);
            this.refresh();
        });

        this.zeroconf.on("remove", (removed: string) => {
            console.log("Zeroconf remove.", removed);
            this.onServiceLost(removed);
            this.refresh();
        });

        this.zeroconf.scan("fk", "tcp", "local.");

        const socket = dgram.createSocket("udp4");
        socket.bind(22143, function () {
            console.log("udp: bound");
            socket.addMembership("224.1.2.3");
        });

        socket.on("message", (buffer, rinfo) => {
            const addresses = [rinfo.address];
            console.log("udp: message", buffer, addresses);
            const decoded = AppProto.UdpMessage.decodeDelimited(buffer);
            const deviceId = Buffer.from(decoded.deviceId).toString("hex");
            console.log("udp: decoded", decoded);
            this.onUdpFound(deviceId, addresses, 80);
        });

        const periodicRefresh = () => {
            setTimeout(() => {
                this.refresh();
                periodicRefresh();
            }, 1000);
        };

        periodicRefresh();

        return Promise.resolve();
    }

    refresh() {
        this.callback(this.getServices());
    }

    onUdpFound(name: string, addresses: string[], port: number) {
        console.log("onUdpFound", name, addresses, port);
        if (!this.services[name]) {
            this.services[name] = new Registration(name, addresses, port);
        }
        this.services[name].udp = new Date();
        this.services[name].lost = null;
    }

    onZeroConfFound(name: string, addresses: string[], port: number) {
        console.log("onZeroConfFound", name, addresses, port);
        if (!this.services[name]) {
            this.services[name] = new Registration(name, addresses, port);
        }
        this.services[name].zeroconf = new Date();
        this.services[name].lost = null;
    }

    onServiceLost(name: string) {
        if (this.services[name]) {
            console.log("onServiceLost", name);
            this.services[name].lost = new Date();
        } else {
            console.log("onServiceLost (MYSTERY)", name);
        }
    }

    getServices(): Registration[] {
        return Object.values(this.services);
    }

    async query(name: string): Promise<void> {
        const service = this.services[name];
        if (!service) {
            console.log("querying (MYSTERY)", name);
            return Promise.resolve();
        }

        console.log("query-service", service);

        const url = `http://${service.addresses[0]}:${service.port}/fk/v1`;

        console.log("query-url", url);

        service.queried = new Date();

        this.refresh();

        await RNFetchBlob.fetch("GET", url)
            .then((res: ResponseType) => {
                service.replied = new Date();
                console.log("response", res.info());
                const status = res.info().status;
                if (status == 200) {
                    const base64Str = res.base64();
                    const buffer = Buffer.from(base64Str, "base64");
                    const decoded = AppProto.HttpReply.decodeDelimited(buffer);
                    console.log("query-decoded", decoded);
                }
            })
            .catch((errorMessage: string, statusCode: number) => {
                console.log(`query-error-`, statusCode, errorMessage);
            });

        this.refresh();

        return Promise.resolve();
    }
}

type ResponseType = {};
