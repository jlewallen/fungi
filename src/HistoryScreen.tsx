import React, { useState } from "react";
import { StatusBar, StyleSheet, Text, FlatList, useColorScheme, View } from "react-native";
import { Colors } from "react-native/Libraries/NewAppScreen";

import { Discovery, HistoryEntry, HistoryEntryType } from "./discovery";

const HistoryItem = ({ item }: { item: HistoryEntry }) => {
    const age = new Date() - item.time;

    const typeToView = (type: HistoryEntryType): string => {
        switch (type) {
            case HistoryEntryType.Query:
                return "Query";
            case HistoryEntryType.UdpMessage:
                return "UdpMessage";
            case HistoryEntryType.ZeroConfFound:
                return "ZeroConfFound";
            case HistoryEntryType.ZeroConfLost:
                return "ZeroConfLost";
        }
        return "Unknown";
    };

    return (
        <View style={{ margin: 10 }}>
            <Text style={{ fontWeight: "800" }}>{typeToView(item.type)}</Text>
            <Text style={{ fontSize: 14 }}>{item.deviceId}</Text>
            <Text style={{ fontSize: 12 }}>{age} ms ago.</Text>
        </View>
    );
};

export const HistoryScreen = ({ discovery, navigation }: { discovery: Discovery; navigation: unknown }) => {
    const isDarkMode = useColorScheme() === "dark";

    const backgroundStyle = {
        backgroundColor: isDarkMode ? Colors.darker : Colors.lighter,
    };

    const [history, setHistory] = useState<HistoryEntry[]>([]);

    React.useEffect(() => discovery.subscribe("history", (history: HistoryEntry[]) => setHistory(history)), []);

    return (
        <View>
            <StatusBar barStyle={isDarkMode ? "light-content" : "dark-content"} />
            <FlatList
                style={backgroundStyle}
                data={history}
                keyExtractor={(item) => item.key}
                renderItem={(row) => <HistoryItem item={row.item} />}
            />
        </View>
    );
};

const styles = StyleSheet.create({
    stationContainer: {
        marginTop: 10,
        paddingHorizontal: 24,
    },
    stationTitle: {
        fontSize: 24,
        fontWeight: "600",
        padding: 5,
    },
    stationId: {
        marginTop: 8,
        fontSize: 14,
        fontWeight: "200",
    },
    stationSeen: {
        marginTop: 8,
        fontSize: 14,
        fontWeight: "400",
    },
    stationDescription: {
        marginTop: 8,
        fontSize: 18,
        fontWeight: "400",
    },
    highlight: {
        fontWeight: "700",
    },
    buttonContainerStyle: {
        marginTop: 10,
    },
});
