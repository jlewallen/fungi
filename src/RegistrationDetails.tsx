import moment from "moment";

import React, { useState } from "react";
import { Button, StyleSheet, Text, useColorScheme, View } from "react-native";
import { Colors } from "react-native/Libraries/NewAppScreen";

import { Discovery, PersistedStation, Registration } from "./discovery";
import { StationNavigation } from "./types";

export const RegistrationDetails: React.FC<{
    registration: Registration;
    showHeader: boolean;
    discovery: Discovery;
    onPress: (registration: Registration) => void;
}> = ({ children, registration, showHeader, discovery, onPress }) => {
    const isDarkMode = useColorScheme() === "dark";

    const primaryColor = registration.lost ? "coral" : "darkseagreen";

    const renderAge = (value: Date | null): string => {
        if (!value) {
            return "Never";
        }
        return moment().diff(value, "seconds") + " seconds ago";
    };

    const sections = [];

    const pushAgeSection = (label: string, value: Date | null) => {
        if (!value) {
            return;
        }
        sections.push(
            <Text
                key={sections.length}
                style={[
                    styles.stationSeen,
                    {
                        color: isDarkMode ? Colors.white : Colors.black,
                    },
                ]}
            >
                {label}: {renderAge(value)}
            </Text>
        );
    };

    if (showHeader) {
        sections.push(
            <Text
                key={sections.length}
                onPress={() => onPress(registration)}
                style={[
                    styles.stationTitle,
                    {
                        backgroundColor: primaryColor,
                        color: isDarkMode ? Colors.white : Colors.black,
                    },
                ]}
            >
                {registration.address}
            </Text>
        );
    }
    if (registration.lost) {
        sections.push(
            <Text
                key={sections.length}
                style={[
                    styles.stationSeen,
                    {
                        color: isDarkMode ? Colors.white : Colors.black,
                    },
                ]}
            >
                LOST: {renderAge(registration.lost)}
            </Text>
        );
    }
    sections.push(
        <Text
            key={sections.length}
            style={[
                styles.stationId,
                {
                    color: isDarkMode ? Colors.white : Colors.black,
                },
            ]}
        >
            {registration.deviceId}
        </Text>
    );

    pushAgeSection("ZeroConf", registration.zeroconf);
    pushAgeSection("UDP", registration.udp);
    pushAgeSection("Queried", registration.queried);
    pushAgeSection("Reply", registration.replied);

    sections.push(
        <View style={styles.buttonContainerStyle} key={sections.length}>
            <Button title="Query" onPress={() => discovery.query(registration.deviceId)} />
        </View>
    );

    return (
        <View style={styles.stationContainer}>
            {sections}
            <Text
                style={[
                    styles.stationDescription,
                    {
                        color: isDarkMode ? Colors.light : Colors.dark,
                    },
                ]}
            >
                {children}
            </Text>
        </View>
    );
};

const styles = StyleSheet.create({
    stationContainer: {
        marginTop: 10,
        paddingHorizontal: 24,
    },
    stationTitle: {
        fontSize: 24,
        fontWeight: "600",
        padding: 5,
    },
    stationId: {
        marginTop: 8,
        fontSize: 14,
        fontWeight: "200",
    },
    stationSeen: {
        marginTop: 8,
        fontSize: 14,
        fontWeight: "400",
    },
    stationDescription: {
        marginTop: 8,
        fontSize: 18,
        fontWeight: "400",
    },
    highlight: {
        fontWeight: "700",
    },
    buttonContainerStyle: {
        marginTop: 10,
    },
});
