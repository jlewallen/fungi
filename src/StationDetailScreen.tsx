import moment from "moment";

import React, { useState } from "react";
import { ScrollView, StatusBar, StyleSheet, Text, useColorScheme, View } from "react-native";
import { Colors } from "react-native/Libraries/NewAppScreen";

import { Discovery } from "./discovery";
import { StationNavigation, PersistedStation, Registration } from "./types";

import { RegistrationDetails } from "./RegistrationDetails";

const StationInspection: React.FC<{ station: PersistedStation }> = ({ children, station }) => {
    console.log("modules", station.reply.modules);
    console.log("live-readings", station.reply.liveReadings);
    return (
        <ScrollView style={{ paddingHorizontal: 24 }}>
            <Text>Name: {station.reply.status?.identity?.name}</Text>
            <Text>Uptime: {station.reply.status?.uptime}</Text>
            <Text>Battery: {station.reply.status?.power?.battery?.voltage}</Text>
            {station.reply.liveReadings.map((lr) => {
                return lr.modules.map((mod, modIndex: number) => {
                    return (
                        <View key={modIndex} style={{ marginBottom: 5, marginTop: 5 }}>
                            <Text style={{ fontSize: 18, fontWeight: "800" }}>{mod.module?.name}</Text>
                            {mod.readings?.map((reading, readingIndex: number) => {
                                return (
                                    <View
                                        key={readingIndex}
                                        style={{
                                            flexDirection: "row",
                                            display: "flex",
                                            justifyContent: "space-evenly",
                                        }}
                                    >
                                        <Text style={{ fontWeight: "600" }}>{reading.sensor?.name}</Text>
                                        <Text>{reading.uncalibrated?.toFixed(3)}</Text>
                                        <Text>{reading.factory?.toFixed(3)}</Text>
                                        <Text>{reading.value?.toFixed(3)}</Text>
                                    </View>
                                );
                            })}
                        </View>
                    );
                });
            })}
            <Text>Firmware: {station.reply.status?.firmware?.version}</Text>
            <Text>Solar: {station.reply.status?.power?.solar?.voltage}</Text>
            <Text>GPS: {station.reply.status?.gps?.fix}</Text>
            <Text>Latitude: {station.reply.status?.gps?.latitude}</Text>
            <Text>Longitude: {station.reply.status?.gps?.longitude}</Text>
            {station.reply.streams.map((stream) => {
                return (
                    <View key={stream.name}>
                        <Text>
                            {stream.name}: {stream.block} blocks, {stream.size} bytes
                        </Text>
                    </View>
                );
            })}
        </ScrollView>
    );
};

export const StationDetailScreen = ({ discovery, navigation, route }: { discovery: Discovery; navigation: unknown; route: unknown }) => {
    const isDarkMode = useColorScheme() === "dark";

    const backgroundStyle = {
        backgroundColor: isDarkMode ? Colors.darker : Colors.lighter,
    };

    const stationNavigation: StationNavigation = route.params.station;

    console.log("station-detail-navigation", stationNavigation);

    const [station, setStation] = useState<PersistedStation | null>(null);
    const [registration, setRegistration] = useState<Registration | null>(null);

    React.useEffect(
        () =>
            discovery.subscribe(`stations/${stationNavigation.deviceId}`, (station: PersistedStation) => {
                console.log("station-detail-station", station);
                console.log("station-detail-station", station.reply.status);
                setStation(station);
            }),
        [stationNavigation.deviceId]
    );

    React.useEffect(
        () =>
            discovery.subscribe(`registrations/${stationNavigation.deviceId}`, (registration: Registration) => {
                console.log("station-detail-registration", registration);
                setRegistration(registration.clone());
            }),
        [stationNavigation.deviceId]
    );

    if (!registration) {
        return (
            <View>
                <StatusBar barStyle={isDarkMode ? "light-content" : "dark-content"} />
                <Text style={{ margin: 20 }}>Loading</Text>
            </View>
        );
    }

    return (
        <View style={{ flex: 1 }}>
            <StatusBar barStyle={isDarkMode ? "light-content" : "dark-content"} />
            <RegistrationDetails registration={registration} showHeader={false} onPress={console.log} discovery={discovery} />
            {station && <StationInspection station={station} />}
        </View>
    );
};

const styles = StyleSheet.create({
    stationContainer: {
        marginTop: 10,
        paddingHorizontal: 24,
    },
    stationTitle: {
        fontSize: 24,
        fontWeight: "600",
        padding: 5,
    },
    stationId: {
        marginTop: 8,
        fontSize: 14,
        fontWeight: "200",
    },
    stationSeen: {
        marginTop: 8,
        fontSize: 14,
        fontWeight: "400",
    },
    stationDescription: {
        marginTop: 8,
        fontSize: 18,
        fontWeight: "400",
    },
    highlight: {
        fontWeight: "700",
    },
    buttonContainerStyle: {
        marginTop: 10,
    },
});
