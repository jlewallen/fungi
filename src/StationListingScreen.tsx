import React, { useState } from "react";
import { StyleSheet, StatusBar, Button, Text, FlatList, useColorScheme, View, Switch } from "react-native";
import { Colors } from "react-native/Libraries/NewAppScreen";

import { Discovery } from "./discovery";
import { StationNavigation, PersistedStation, Registration } from "./types";

import { RegistrationDetails } from "./RegistrationDetails";

export const StationListingScreen = ({ discovery, navigation }: { discovery: Discovery; navigation: unknown }) => {
    const isDarkMode = useColorScheme() === "dark";

    const backgroundStyle = {
        backgroundColor: isDarkMode ? Colors.darker : Colors.lighter,
    };

    const [isPassive, setIsPassive] = useState<boolean>(discovery.passive);
    const [registrations, setRegistrations] = useState<Registration[]>([]);

    const togglePassive = () => {
        setIsPassive(!isPassive);
        discovery.setPassive(!isPassive);
    };

    React.useEffect(() => discovery.subscribe("registrations", (registrations: Registration[]) => setRegistrations(registrations)), []);

    const onSelected = (registration: Registration) => {
        navigation.navigate("StationDetail", { station: StationNavigation.fromRegistration(registration) });
    };

    const onHistory = () => {
        navigation.navigate("History");
    };

    return (
        <View>
            <StatusBar barStyle={isDarkMode ? "light-content" : "dark-content"} />
            <View
                style={{
                    margin: 20,
                    flexDirection: "row",
                    justifyContent: "flex-start",
                    alignItems: "center",
                    ...backgroundStyle,
                }}
            >
                <Switch
                    accessibilityLabel="Query Stations."
                    trackColor={{ false: "#767577", true: "#81b0ff" }}
                    thumbColor={isPassive ? "#f5dd4b" : "#f4f3f4"}
                    ios_backgroundColor="#3e3e3e"
                    onValueChange={togglePassive}
                    value={!isPassive}
                />
                <Text style={{ marginLeft: 10, fontWeight: "800" }}>Query Stations ({isPassive ? "No" : "Yes"})</Text>
            </View>
            <View
                style={{
                    margin: 20,
                    ...backgroundStyle,
                }}
            >
                <Button title="History" onPress={() => onHistory()} />
            </View>
            <FlatList
                style={backgroundStyle}
                data={registrations}
                keyExtractor={(item) => item.deviceId}
                renderItem={(row) => (
                    <RegistrationDetails onPress={onSelected} registration={row.item} showHeader={true} discovery={discovery} />
                )}
            />
        </View>
    );
};

const styles = StyleSheet.create({
    stationContainer: {
        marginTop: 10,
        paddingHorizontal: 24,
    },
    stationTitle: {
        fontSize: 24,
        fontWeight: "600",
        padding: 5,
    },
    stationId: {
        marginTop: 8,
        fontSize: 14,
        fontWeight: "200",
    },
    stationSeen: {
        marginTop: 8,
        fontSize: 14,
        fontWeight: "400",
    },
    stationDescription: {
        marginTop: 8,
        fontSize: 18,
        fontWeight: "400",
    },
    highlight: {
        fontWeight: "700",
    },
    buttonContainerStyle: {
        marginTop: 10,
    },
});
