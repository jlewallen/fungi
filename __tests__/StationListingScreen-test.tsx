/**
 * @format
 */

import "react-native";
import React from "react";
// Note: test renderer must be required after react-native.
import renderer from "react-test-renderer";
import EventEmitter from "tiny-emitter";

jest.mock("react-native/Libraries/Components/Switch/Switch", () => {
    const mockComponent = require("react-native/jest/mockComponent");
    return mockComponent("react-native/Libraries/Components/Switch/Switch");
});

import { StationListingScreen } from "../src/StationListingScreen";

const fakeDiscovery = new EventEmitter();

it("renders correctly", async () => {
    const root = renderer.create(
        <StationListingScreen discovery={fakeDiscovery} navigation={{ navigate: () => console.log("navigate") }} />
    );

    expect(root.toJSON()).toMatchSnapshot();
});
