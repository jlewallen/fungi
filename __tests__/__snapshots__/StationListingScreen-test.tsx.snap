// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F3F3F3",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "margin": 20,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text>
        true
      </Text>
    </View>
    <Text
      style={
        Object {
          "fontWeight": "800",
          "marginLeft": 10,
        }
      }
    >
      Query Stations (
      Yes
      )
    </Text>
  </View>
  <RCTScrollView
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "backgroundColor": "#F3F3F3",
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</View>
`;
